cassiopeia/solver/DistanceSolver.py

```diff
@@ -167,42 +125,62 @@ def solve(self):
         tree.add_edge(remaining_samples[0], remaining_samples[1])
 
         tree = nx.relabel_nodes(tree, identifier_to_sample)
-        tree = self.append_sample_names(tree)
-        tree = self.root_tree(tree)
 
-        self.tree = tree
+        if cassiopeia_tree.root_sample_name is not None:
+            tree = self.root_tree(tree, cassiopeia_tree.root_sample_name)
 
-    @numba.jit(forceobj=True)
-    def compute_dissimilarity_map(self, cm: np.array, C: int) -> np.array:
-        """Compute the dissimilarity between all samples
+        cassiopeia_tree.populate_tree(tree)
 
-        An optimized function for computing pairwise dissimilarities between
-        samples in a character matrix according to the dissimilarity function.
+    def setup_dissimilarity_map(
+        self, cassiopeia_tree: CassiopeiaTree
+    ) -> Tuple[pd.DataFrame, pd.DataFrame, Dict[str, List[str]], str]:
```

Comment:
I find the logic behind setting up the dissimilarity map tricky: Suppose the `cassiopeia_tree` already has a dissimilarity map in it, and the solver was initialized with a `dissimilarity_function`. Which of the two dissimilarity maps will prevail, and why? Might this surprise users?

Reply:
The issue here is if the user wants to root the tree and has not specified a root sample -- then, we'll have to _add_ the root sample to the character matrix and recompute the dissimilarities according to the user-defined dissimilarity map. 

If the user _has_ a root sample, or does not desire to get a rooted tree, then we're good to go and nothing needs to be done (the dissimilarity map will not be overwritten and will be used as is).

Reply:
I see, there are some details involved that exceed my knowledge of the solver. I am cool with this, just wanted to make sure the logic is what we want!